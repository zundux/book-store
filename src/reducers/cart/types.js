export const ADD_TO_CART = "cart/ADD_TO_CART";
export const REMOVE_FROM_CART = "cart/REMOVE_FROM_CART";
export const CHANGE_PRODUCT_AMOUNT_IN_CART = "cart/CHANGE_PRODUCT_AMOUNT_IN_CART";
export const CLEAN_CART = "cart/CLEAN_CART";
