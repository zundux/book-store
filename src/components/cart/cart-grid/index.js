import CartGrid from "./cart-grid";

export default CartGrid;