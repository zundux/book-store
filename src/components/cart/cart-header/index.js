import CartHeader from "./cart-header";

export default CartHeader;