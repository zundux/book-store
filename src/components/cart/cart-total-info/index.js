import CartTotalInfo from "./cart-total-info";

export default CartTotalInfo;