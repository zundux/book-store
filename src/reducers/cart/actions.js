import * as R from "ramda";
import { ADD_TO_CART, CHANGE_PRODUCT_AMOUNT_IN_CART, CLEAN_CART, REMOVE_FROM_CART } from "./types";

export const addToCart = (product, amount) => dispatch => dispatch({
  type: ADD_TO_CART,
  product,
  amount
});

export const removeFromCart = (product) => dispatch => dispatch({
  type: REMOVE_FROM_CART,
  product
});

export const cleanCart = () => dispatch => dispatch({
  type: CLEAN_CART
});

export const changeAmountOfProductInCart = (product, amount) => dispatch => {
  if (!R.is(Number, amount) || isNaN(amount) || amount <= 0) {
    amount = 1;
  }

  dispatch({
    type: CHANGE_PRODUCT_AMOUNT_IN_CART,
    product,
    amount
  });
};

