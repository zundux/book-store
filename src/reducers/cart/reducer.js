import { ADD_TO_CART, CHANGE_PRODUCT_AMOUNT_IN_CART, CLEAN_CART, REMOVE_FROM_CART } from "./types";

export const initialStateCart = {
  products: []
};

const byId = targetItem => {
  return item => item.id === targetItem.id;
};

const getExistListItemIndex = (list, targetItem) => list.findIndex(byId(targetItem));

const productReducer = (state = undefined, action) => {
  switch (action.type) {
    case ADD_TO_CART:
      let amount;

      if (state) {
        if (action.amount) {
          amount = state.amount + action.amount;
        } else {
          amount = state.amount + 1;
        }
      } else {
        if (action.amount) {
          amount = action.amount;
        } else {
          amount = 1;
        }
      }

      return {
        ...action.product,
        amount
      };


    case CHANGE_PRODUCT_AMOUNT_IN_CART:
      return {
        ...state,
        amount: action.amount || 0
      };


    default:
      return state;
  }
};


const cartReducer = (state = initialStateCart, action) => {
  let existProductIndex;
  let isNotExistProduct;

  if (state.products.length && action.product) {
    existProductIndex = getExistListItemIndex(state.products, action.product);
    isNotExistProduct = existProductIndex === -1;
  }

  switch (action.type) {
    case ADD_TO_CART: {
      if (isNotExistProduct) {

        return {
          products: [
            ...state.products,
            productReducer(undefined, action)
          ]
        };
      }

      const product = state.products[existProductIndex];

      return {
        products: [
          ...state.products.slice(0, existProductIndex),
          productReducer(product, action),
          ...state.products.slice(existProductIndex + 1)
        ]
      };
    }

    case REMOVE_FROM_CART:
      if (isNotExistProduct) {
        return state;
      }

      return {
        products: [
          ...state.products.slice(0, existProductIndex),
          ...state.products.slice(existProductIndex + 1)
        ]
      };

    case CHANGE_PRODUCT_AMOUNT_IN_CART:
      if (isNotExistProduct) {
        return state;
      }

      return {
        products: [
          ...state.products.slice(0, existProductIndex),
          productReducer(state.products[existProductIndex], action),
          ...state.products.slice(existProductIndex + 1)
        ]
      };

    case CLEAN_CART:
      return { ...initialStateCart };

    default:
      return state;
  }
};

export default cartReducer;