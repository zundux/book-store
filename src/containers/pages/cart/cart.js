import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import CartHeader from "../../../components/cart/cart-header";
import CartGrid from "../../../components/cart/cart-grid";
import { changeAmountOfProductInCart, removeFromCart } from "../../../reducers/cart";
import CartTotalInfo from "../../../components/cart/cart-total-info";
import { Link } from "react-router-dom";
import * as R from "ramda";

export const Cart = ({ products, onRemoveFromCart, onChangeAmountOfProductInCart }) => {
  const isNotEmpty = !R.either(R.isNil, R.isEmpty)(products);

  return (
    <div>
      <CartHeader products={products}/>
      <CartGrid {...{
        products,
        onRemoveFromCart,
        onChangeAmountOfProductInCart
      }}/>
      <CartTotalInfo {...{ products }}/>
      {isNotEmpty && (
        <button>
          <Link to="/order">Checkout</Link>
        </button>
      )}
    </div>
  );
};

const mapStateToProps = ({ cart }) => ({
  products: cart.products
});

const mapDispatchToProps = dispatch => bindActionCreators({
  onRemoveFromCart: removeFromCart,
  onChangeAmountOfProductInCart: changeAmountOfProductInCart
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Cart);