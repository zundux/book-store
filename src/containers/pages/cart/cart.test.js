import React from "react";
import { Cart } from "./cart";

describe("Cart page", () => {

  it("Cart page should not show checkout link when cart is empty", () => {
    const products = [];
    const onRemoveFromCart = () => {
    };
    const onChangeAmountOfProductInCart = () => {
    };

    const component = shallow(
      <Cart {...{
        products,
        onRemoveFromCart,
        onChangeAmountOfProductInCart
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
    // todo should NOT have the checkout button
  });

  it("Cart page should show the checkout link when cart has some products or one", () => {
    const products = [{ id: 1 }, { id: 2 }, { id: 3 }];
    const onRemoveFromCart = () => {
    };
    const onChangeAmountOfProductInCart = () => {
    };

    const component = shallow(
      <Cart {...{
        products,
        onRemoveFromCart,
        onChangeAmountOfProductInCart
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
    // todo should HAVE the checkout button
  });
});