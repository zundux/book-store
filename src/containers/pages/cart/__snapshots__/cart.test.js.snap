// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart page Cart page should not show checkout link when cart is empty 1`] = `
<div>
  <CartHeader
    products={Array []}
  />
  <CartGrid
    onChangeAmountOfProductInCart={[Function]}
    onRemoveFromCart={[Function]}
    products={Array []}
  />
  <CartTotalInfo
    products={Array []}
  />
</div>
`;

exports[`Cart page Cart page should show the checkout link when cart has some products or one 1`] = `
<div>
  <CartHeader
    products={
      Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
      ]
    }
  />
  <CartGrid
    onChangeAmountOfProductInCart={[Function]}
    onRemoveFromCart={[Function]}
    products={
      Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
      ]
    }
  />
  <CartTotalInfo
    products={
      Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
        Object {
          "id": 3,
        },
      ]
    }
  />
  <button>
    <Link
      replace={false}
      to="/order"
    >
      Checkout
    </Link>
  </button>
</div>
`;
